pkg/flowaggregator/certificate/provider.go

```diff
@@ -0,0 +1,623 @@
+	}
+
+	if err = p.syncCertificateSecrets(ctx, caSecretName, caCertPEM, caKeyPEM, caSecret); err != nil {
+		return nil, nil, fmt.Errorf("failed to sync CA secret: %w", err)
+	}
+
+	klog.Info("CA certificate rotated")
+
+	return caCertPEM, caKeyPEM, nil
+}
+
+func (p *Provider) rotateClientCertificate(ctx context.Context, caCertPEM []byte, caKeyPEM []byte) error {
+	clientCertPEM, _, clientSecret, err := p.getSecret(clientSecretName)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to get client secret: %w", err)
+		}
+	}
+
+	if err := verifyCertificate(caCertPEM, clientCertPEM, p.clock.Now()); err != nil {
+		// Client cert could not be validated by the CA, regenerate client based on CA.
+		caCert, caKey, err := pemToCertKey(caCertPEM, caKeyPEM)
+		if err != nil {
+			return err
+		}
+
+		clientCertPEM, clientKeyPEM, err := GenerateCertKey(caCert, caKey, p.validFrom(), false, "")
+		if err != nil {
+			return fmt.Errorf("failed to generate client cert: %w", err)
+		}
+		if err := p.syncCertificateSecrets(ctx, clientSecretName, clientCertPEM, clientKeyPEM, clientSecret); err != nil {
+			if !errors.IsAlreadyExists(err) && !errors.IsConflict(err) {
+				return fmt.Errorf("failed to sync client Secret: %w", err)
+			}
+		}
+
+		klog.Info("Client certificates were rotated")
+	}
+	return nil
+}
+
+func (p *Provider) validFrom() time.Time {
+	// We set certificates as valid starting a hour earlier due to potential time skew on nodes.
+	return p.clock.Now().Add(-time.Hour)
+}
+
+func (p *Provider) shouldRotateCertificate(certPEM []byte, keyPEM []byte) bool {
+	if len(certPEM) == 0 || len(keyPEM) == 0 {
+		return true
+	}
+
+	cert, err := tls.X509KeyPair(certPEM, keyPEM)
+	if err != nil {
+		klog.ErrorS(err, "Certificate not signed by the correct private key")
+		return true
+	}
+
+	remainingDuration := cert.Leaf.NotAfter.Sub(p.clock.Now())
+	if remainingDuration < minValidDuration {
+		klog.InfoS("The remaining duration of the TLS certificate and key is less than min valid duration", "remaining", remainingDuration, "min", minValidDuration)
+		return true
+	}
+	return false
+}
+
+func (p *Provider) getSecret(name string) ([]byte, []byte, *v1.Secret, error) {
+	var obj any
+	var exists bool
+	var err error
+	switch name {
+	case caSecretName:
+		obj, exists, err = p.caSecretInformer.GetIndexer().GetByKey(fmt.Sprintf("%s/%s", p.namespace, caSecretName))
+	case clientSecretName:
+		obj, exists, err = p.clientSecretInformer.GetIndexer().GetByKey(fmt.Sprintf("%s/%s", p.namespace, clientSecretName))
+	default:
+		return nil, nil, nil, fmt.Errorf("secret %q not managed by flow aggregator", name)
+	}
+
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("error getting Secret %q: %w", name, err)
+	}
+
+	if !exists {
+		return nil, nil, nil, errors.NewNotFound(v1.Resource("secret"), name)
+	}
+
+	secret := obj.(*v1.Secret)
+	return secret.Data[v1.TLSCertKey], secret.Data[v1.TLSPrivateKeyKey], secret.DeepCopy(), nil
```

Comment:
can you let other functions make their own deep copy as needed, it will be more readable?

Reply:
Done